### DISCUSSION TITLE
Is there a way to add extra `options` after a type has been created

### DISCUSSION THREAD

**Discussion by bhuynhdev** [sinclairzx81/typebox#697] [Q&A] [answered]
I wish to add extra `options` parameters to a type after it has been created, such as `examples` or `format`.
```ts
import { Type as t } from '@sinclair/typebox'

const S = t.String()

const S1 = t.MagicalExtend(S, { examples: ['home address']})

const S2 = t.MagicalExtend(S, {examples: ['date string']})
```
Is there a way to achieve this `MagicalExtend` (or perhaps `MagicalRecreate`) feature?

**Comment by sinclairzx81** [comment, marked as answer]
@bhuynhdev Hi,

TypeBox types just Json Schema objects, so you can just apply additional properties using the JavaScript spread operator. Below applies additional properties from T1 to T2, where we use `typeof T1` to give a correct TSchema type of T2 (as they should be the same type)

```typescript
import { Type as t } from '@sinclair/typebox'

const T1 = t.String()                       // const T1: TString = { type: 'string' }

const T2: typeof T1 = { foo: 'bar', ...T1 } // const T2: TString = { type: 'string', foo: 'bar' }
```
So, the above applies to all TB types. You may wish to implement a function to make application of additional properties more consistent (where the options are applied before the schema to ensure no properties are overwritten).
```typescript
import { Type as t, TSchema } from '@sinclair/typebox'

export function MagicExtend<T extends TSchema>(schema: T, options: Record<PropertyKey, unknown>): T {
  return { ...options, ...schema }
}

const T1 = t.String()                       // const T1: TString = { type: 'string' }

const T2 = MagicExtend(T1, { foo: 'bar' })  // const T2: TString = { type: 'string', foo: 'bar' }
```
Hope this helps
S

**Reply by bhuynhdev** [reply]
@sinclairzx81 Hi. Thanks a lot for the lightning fast response. This is awesome. As a follow up on your solution, is there a way to get Typescript to infer the keys in the `options` parameter? For example, if `T` is `Type.Number` then `options` will autocomplete suggest `maximum`, `minimum`, whereas if `T` is `Type.String` then `options` will suggest `format` or `minLength`

Edit: Digging through the codebase, I found multiple kinds of options, so I wrote this. Unsure if this is correct.

```ts
type DynamicTypeboxSchemaOptions<T extends TSchema> = T extends TString
	? StringOptions
	: T extends TArray
	? ArrayOptions
	: T extends TNumber | TInteger
	? NumericOptions<number>
	: T extends TIntersect
	? IntersectOptions
	: T extends TObject | TPartial<T>
	? ObjectOptions
	: SchemaOptions

export function extendSchema<T extends TSchema>(schema: T, options: DynamicTypeboxSchemaOptions<T>): T {
	return { ...options, ...schema }
}
```

TComposite and TRecord did not have any special handling, as I deemed it to be a bit difficult, and my use case would never result in TComposite being passed to `extendSchema` anyway. (Use case: I am using Drizzle ORM's [`drizzle-typebox`](https://orm.drizzle.team/docs/typebox) to `createSelectSchema` from Drizzle's table definition, but I sometimes need to extend the schema field to add `examples` and `format` and `minLength`)

**Reply by sinclairzx81** [reply]
@bhuynhdev No problem. You can try the following which should generalize across all types.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIY64ykzD5wC+cAZlBCHAOQACAzsAHb4ANqWBQA9DHQYARhAAeHAFBKpmOABFU-UiFppMc+QWJkA8mBoR+PADzI4GeTAz8AJjzxES5OGcfOrh5wAEoY+NButgAKrJiwqADSGKg4AK78ANb8EADu-AB8BXAAvChwAGR+Kk6QsMwZ+Fb8AS7uJj72rUGeyB1kOP5ObcFhEVBRsRDxUsmpcBnZeYWliPQFABQ83mQAXCg40808+2YAlPtaOnr4BrIK-aQWx-aDxQhKAJBQGDBpUC0kAA6EFHYDWHg4EFA7amUgMJT0FQRGzwBxlO5AswyABW4RgGw+cGJcHk+0xADk0iAZBgoBszlglCS4KhydIgZQoAIAOYMpksgBe7MwQIAQhAIEIMKR+AzEWdkRD4ABRVbDIKPDbIHBIJiS-ZIGSkKCGuDG4VwACMDFtjFInhRPBgiqUTvgFNWKqB+ogQONUH9pEFJPE4jg7rgFP2VpUagwcDVZUo1FotnjECYieKYbgEEyQA)

```typescript
import { Type, TSchema, Static } from '@sinclair/typebox'

type DynamicTypeboxSchemaOptions<T extends TSchema, O extends Record<PropertyKey, unknown>> = T & O

export function extendSchema<T extends TSchema, O extends Record<PropertyKey, unknown> = {}>(schema: T, options: O): DynamicTypeboxSchemaOptions<T, O> {
	return { ...options, ...schema }
}

const T = Type.Object({
    x: Type.Number(),
    y: Type.String(),
    z: Type.Boolean()
})

const E = extendSchema(T, { foo: { bar: { baz: 1 } } } as const)

const N = E.foo.bar.baz   // const N: 1

type E = Static<typeof E> // ok
```
